File: app/src/main/java/com/maple/audiometry/utils/SinWave.java
package com.maple.audiometry.utils;

public class SinWave {
	/** 正弦波的高度 **/
	public static double HEIGHT = 127;
	/** 2PI **/
	public static final double TWOPI = 2 * 3.1415;

	/**
	 * 生成正弦波
	 * 
	 * @param waveLen
	 *            每段正弦波的长度
	 * @param length
	 *            总长度
	 * @return
	 */
	public static byte[] sin(int waveLen, int length) {
		byte[] wave = new byte[length];
		for (int i = 0; i < length; i++) {
			wave[i] = (byte) (HEIGHT * (1 - Math.sin(TWOPI
					* ((i % waveLen) * 1.00 / waveLen))));
		}
		return wave;
	}

	/**
	 * 更新声音的分贝
	 * 
	 * @param hz
	 * @param dB
	 */
	public static void updateDB(int hz, int dB) {
		double temp = 0;
		switch (hz) {
		case 1000:// 1000频率
			temp = 7;
			break;
		case 2000:// 2000频率
			temp = 9;
			break;
		case 4000:// 4000频率
			temp = 9.5;
			break;
		case 8000:// 8000频率
			temp = 13;
			break;
		case 500:// 500频率
			temp = 11.5;
			break;
		case 250:// 250频率
			temp = 25.5;
			break;
		}
		HEIGHT = Math.pow(10, (dB - temp) / 20);
	}
}

File: app/src/main/java/com/maple/audiometry/utils/SinWave.kt
package com.maple.audiometry.utils

object SinWave {
    /** 正弦波的高度  */
    var HEIGHT = 127.0
    /** 2PI  */
    const val TWOPI = 2 * 3.1415

    /**
     * 生成正弦波
     *
     * @param waveLen   每段正弦波的长度
     * @param length    总长度
     * @return
     */
    @JvmStatic
    fun sin(waveLen: Int, length: Int): ByteArray {
        val wave = ByteArray(length)
        for (i in 0 until length) {
            wave[i] = (HEIGHT * (1 - Math.sin(TWOPI * (i % waveLen * 1.00 / waveLen)))).toByte()
        }
        return wave
    }

    /**
     * 更新声音的分贝
     *
     * @param hz
     * @param dB
     */
    @JvmStatic
    fun updateDB(hz: Int, dB: Int) {
        var temp = 0.0
        when (hz) {
            1000 -> temp = 7.0// 1000频率
            2000 -> temp = 9.0// 2000频率
            4000 -> temp = 9.5// 4000频率
            8000 -> temp = 13.0// 8000频率
            500 -> temp = 11.5// 500频率
            250 -> temp = 25.5// 250频率
        }
        HEIGHT = Math.pow(10.0, (dB - temp) / 20)
    }
}
